### PR TITLE
Migrate to ESLint AirBnB rules

### PR DESCRIPTION
**Pull request by Tristan-WorkGH**

### CODE REVIEW

File: src/components/dialogs/commons/upload-new-case.tsx [hunk 1]
```diff
@@ -124,6 +127,7 @@ const UploadNewCase: React.FunctionComponent<UploadNewCaseProps> = ({
             </Grid>
             <Grid item sx={{ fontWeight: 'bold' }}>
                 <p>
+                    {/* eslint-disable-next-line no-nested-ternary -- "if" can't be used inside jsx block */}
                     {caseFileLoading ? (
                         <CircularProgress size="1rem" />
                     ) : caseFileName ? (
```
Comment (Meklo) [inline src/components/dialogs/commons/upload-new-case.tsx]: How about externalizing the nested ternary in a render function with if else statements ?

File: src/components/dialogs/contingency-list/contingency-list-utils.ts [hunk 2]
```diff
@@ -36,6 +36,7 @@ export interface CriteriaBasedData {
     [FieldConstants.NOMINAL_VOLTAGE_1]?: RangeInputData | null;
     [FieldConstants.NOMINAL_VOLTAGE_2]?: RangeInputData | null;
     [FieldConstants.NOMINAL_VOLTAGE_3]?: RangeInputData | null;
+
```
Comment (Meklo) [inline src/components/dialogs/contingency-list/contingency-list-utils.ts]: ?

File: src/module-emotion.d.ts [hunk 3]
```diff
@@ -4,6 +4,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
+/* eslint-disable spaced-comment */
```
Comment (Meklo) [inline src/module-emotion.d.ts:7]: How about removing the comment ?

File: src/components/dialogs/create-case-dialog/create-case-dialog.tsx [hunk 4]
```diff
@@ -82,7 +81,7 @@ const CreateCaseDialog: React.FunctionComponent<CreateCaseDialogProps> = ({ onCl
             .then(onClose)
             .catch((err) => {
                 if (handleMaxElementsExceededError(err, snackError)) {
-                    return;
+                    // do nothing
                 } else if (err?.status === HTTP_UNPROCESSABLE_ENTITY_STATUS) {
```
Comment (Meklo) [inline src/components/dialogs/create-case-dialog/create-case-dialog.tsx]: How about incorporating `!handleMaxElementsExceededError` to the second if else statement ?

File: src/components/dialogs/create-study-dialog/create-study-dialog-utils.ts [hunk 5]
```diff
@@ -6,39 +6,40 @@
...
 
 import { ElementAttributes, FieldConstants, Parameter, yupConfig as yup } from '@gridsuite/commons-ui';
+import { UUID } from 'crypto';
 
 export const getCreateStudyDialogFormDefaultValues = ({
+    // @ts-expect-error how react-hook manage strings like UUIDs?
     directory = '',
     studyName = '',
     caseFile = null,
+    // @ts-expect-error how react-hook manage strings like UUIDs?
```
Comment (Meklo) [inline src/components/dialogs/create-study-dialog/create-study-dialog-utils.ts]: One way could be using `randomUUID()` from `crypto`

File: src/components/directory-breadcrumbs.tsx [hunk 6]
```diff
@@ -70,14 +65,14 @@ const styles = {
     },
 };
 
-const DirectoryBreadcrumbs = () => {
+export default function DirectoryBreadcrumbs() {
     const dispatch = useDispatch();
 
     const selectedDirectory = useSelector((state: AppState) => state.selectedDirectory);
-    const currentPath = useSelector((state: AppState) => state.currentPath);
+    const currentPath = useSelector((state: AppState) => state.currentPath as any[]);
```
Comment (Meklo) [inline src/components/directory-breadcrumbs.tsx]: Shouldn't it be typed as `ElementAttributes[]` ?

File: src/components/directory-tree-view.tsx [hunk 7]
```diff
@@ -68,26 +64,31 @@ const styles = {
...
+    onContextMenu,
+    onDirectoryUpdate,
+}: Readonly<DirectoryTreeViewProps>) {
     const dispatch = useDispatch();
 
-    const [expanded, setExpanded] = React.useState<UUID[]>([]);
+    const [expanded, setExpanded] = useState<UUID[]>([]);
     const selectedDirectory = useSelector((state: AppState) => state.selectedDirectory);
-    const currentPath = useSelector((state: AppState) => state.currentPath);
+    const currentPath = useSelector((state: AppState) => state.currentPath as any[]);
```
Comment (Meklo) [inline src/components/directory-tree-view.tsx]: Shouldn't it be typed as `ElementAttributes[]` ?

File: src/components/menus/common-contextual-menu.tsx [hunk 8]
```diff
@@ -73,19 +76,22 @@ const CommonContextualMenu: React.FC<CommonContextualMenuProps> = (props) => {
...
-                } else {
-                    return makeMenuItem(
-                        index,
-                        menuItem.messageDescriptorId,
-                        menuItem.callback,
-                        menuItem.icon,
-                        menuItem.disabled
+                    return (
+                        <Divider
+                            key={`${menuItem.isDivider ? 'divider' : 'menu'}-${
```
Comment (dbraquart) [inline src/components/menus/common-contextual-menu.tsx]: why changing the key value ?

File: src/components/custom-tree-item.tsx [hunk 9]
```diff
@@ -5,69 +5,105 @@
...
+        expanded?: SxProps<Theme>;
+        selected?: SxProps<Theme>;
+        focused?: SxProps<Theme>;
+        disabled?: SxProps<Theme>;
+        label?: SxProps<Theme>;
+        iconContainer?: SxProps<Theme>;
+    };
+    onExpand: (nodeId: UUID) => void;
+    onSelect: (nodeId: UUID) => void;
 }
```
Comment (dbraquart) [inline src/components/custom-tree-item.tsx:41]: where you forced by the linter to make so many changes in this file ?

File: src/components/dialogs/export-case-dialog.tsx [hunk 10]
```diff
@@ -91,15 +91,25 @@ const ExportCaseDialog = ({ selectedElements, onClose, onExport }: ExportCaseDia
...
-            //TODO to be removed when extensions param default value corrected in backend to include all possible values
-            Object.values(fetchedFormats).forEach((format) =>
-                format.parameters.forEach((param) => {
-                    if (param.type === 'STRING_LIST' && param.name.endsWith('extensions')) {
-                        param.defaultValue = param.possibleValues;
-                    }
-                })
+            // TODO to be removed when extensions param default value corrected in backend to include all possible values
+            setFormats(
+                Array.isArray(fetchedFormats)
```
Comment (dbraquart) [inline src/components/dialogs/export-case-dialog.tsx]: I dont understand the use of isArray here

File: src/utils/custom-hooks.ts [hunk 11]
```diff
@@ -66,13 +65,13 @@ interface MultipleAction<T> {
...
-    onSuccess?: (data: T, args: unknown[]) => void,
+export const useDeferredFetch = <T, TArgs extends unknown[] = any[]>(
+    fetchFunction: GenericFunction<T, TArgs>,
+    onSuccess?: (data: T, args: TArgs) => void,
     errorToString?: (status: number) => string | undefined,
-    onError?: (errorMessage: string, args: unknown[]) => void,
+    onError?: (errorMessage: string, args: TArgs) => void,
     hasResult: boolean = true
-): [(...args: unknown[]) => void, FetchState<T>] => {
+): [(...args: TArgs) => void, FetchState<T>] => {
```
Comment (Meklo) [inline src/utils/custom-hooks.ts:74]: Is this change really worth ? It seems it brings a lot of ts-expect-error annotations

File: src/utils/custom-hooks.ts [hunk 12]
```diff
@@ -158,27 +156,22 @@ export const useDeferredFetch = <T>(
...
-        },
-        [fetchData]
-    );
+    const fetchCallback = useCallback((...args: TArgs) => fetchData(...args), [fetchData]);
 
     return [fetchCallback, state];
 };
 
-/////////////////////////////////////////////////////////////////:
+/// //////////////////////////////////////////////////////////////:
```
Comment (Meklo) [inline src/utils/custom-hooks.ts]: ?

File: src/components/search/search-bar.tsx [hunk 13]
```diff
@@ -15,25 +14,24 @@ import {
...
-export const SEARCH_FETCH_TIMEOUT_MILLIS = 1000; // 1 second
-
-interface SearchBarProps {
+export interface SearchBarProps {
     inputRef: RefObject<TextFieldProps>;
 }
 
-//TODO remove when ElementSearchInputProps is exported in commons-ui
+// TODO remove when ElementSearchInputProps is exported in commons-ui
 type ElementSearchInputProps<T> = Parameters<typeof ElementSearchInput<T>>[0];
```
Comment (Meklo) [inline src/components/search/search-bar.tsx]: I reckon it's now available from commons ui

File: src/components/utils/directory-content-utils.ts [hunk 14]
```diff
@@ -28,28 +28,23 @@ export const formatMetadata = (
...
     !checkedRows.find((checkedRow) => checkedRow.elementUuid === row.elementUuid);
 
-export const defaultColumnDefinition = {
+export const defaultColumnDefinition: ColDef<unknown> = {
     sortable: true,
     resizable: true,
     lockPinned: true,
     wrapHeaderText: true,
     autoHeaderHeight: true,
-    suppressHorizontalScroll: true,
```
Comment (Meklo) [inline src/components/utils/directory-content-utils.ts:52]: Why this removal ?

File: src/components/search/search-item.tsx [hunk 15]
```diff
@@ -32,24 +30,27 @@ const styles = {
...
     const escapedHighlight = highlight.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
     const parts = text.split(new RegExp(`(${escapedHighlight})`, 'gi'));
     return (
         <span>
-            {parts.map((part, i) =>
+            {parts.map((part) =>
                 part.toLowerCase() === highlight.toLowerCase() ? (
-                    <span key={i} style={{ fontWeight: 'bold' }}>
+                    <span
+                        key={`part-${part.toLowerCase().replace(/[^[a-z]\d.-]+/g, '-')}`}
```
Comment (dbraquart) [inline src/components/search/search-item.tsx]: why changing the key here ?

File: src/translations/dynamic/network-modifications-locale-en.ts [hunk 16]
```diff
@@ -5,7 +5,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
-const network_modifications_locale_en = {
+const NetworkModificationsLocaleEn = {
```
Comment (dbraquart) [inline src/translations/dynamic/network-modifications-locale-en.ts:8]: This file is useless. Will create a US for that.

File: src/translations/dynamic/network-modifications-locale-fr.ts [hunk 17]
```diff
@@ -5,7 +5,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
-const network_modifications_locale_fr = {
+const NetworkModificationsLocaleFr = {
```
Comment (dbraquart) [inline src/translations/dynamic/network-modifications-locale-fr.ts:8]: This file is useless. Will create a US for that.

**Review by Meklo — CHANGES_REQUESTED** [review]
There's a new warning when opening directory contextual menu : 
`Warning: Encountered two children with the same key, .$divider-no_msg. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.`

Also I reckon some of the new Sonar issues are easily fixable (using nullish coalescing operator, marking props as read only...)

**Review by dbraquart — COMMENTED** [review]
code: ok - just a few questions
tests: still on-going

**Review by FranckLecuyer — COMMENTED** [review]
Tests: OK

File: src/components/dialogs/commons/upload-new-case.tsx (hunk 1 above)
Comment (Tristan-WorkGH) [reply src/components/dialogs/commons/upload-new-case.tsx]: Found a way around with a `useMemo`.

File: src/module-emotion.d.ts (hunk 3 above)
Comment (Tristan-WorkGH) [reply src/module-emotion.d.ts:7]: It's present on commons-ui as reference for these particular files.

File: src/components/directory-tree-view.tsx (hunk 7 above)
Comment (Tristan-WorkGH) [reply src/components/directory-tree-view.tsx]: Dunno, in the redux definition it's done as `unknown[]` which pose problems of types after.

File: src/components/directory-breadcrumbs.tsx (hunk 6 above)
Comment (Tristan-WorkGH) [reply src/components/directory-breadcrumbs.tsx]: https://github.com/gridsuite/gridexplore-app/pull/546#discussion_r1816172547

File: src/utils/custom-hooks.ts (hunk 11 above)
Comment (Tristan-WorkGH) [reply src/utils/custom-hooks.ts:74]: Yes, because now typescript can infer what the function return.

File: src/utils/custom-hooks.ts (hunk 12 above)
Comment (Tristan-WorkGH) [reply src/utils/custom-hooks.ts]: New rule:
> ESLint: Expected exception block, space or tab after '//' in comment.(spaced-comment)

File: src/components/utils/directory-content-utils.ts (hunk 14 above)
Comment (Tristan-WorkGH) [reply src/components/utils/directory-content-utils.ts:52]: TypeScript error: this parameter does not exist in AgGrid, adding `ColDef` type help typescript to confirm that.

File: src/components/dialogs/create-study-dialog/create-study-dialog-utils.ts (hunk 5 above)
Comment (Tristan-WorkGH) [reply src/components/dialogs/create-study-dialog/create-study-dialog-utils.ts]: @klesaulnier has explain to me that the problem we have is that react and typescript consider `undefined` for not defined values, and react-hook-form consider `null` for them...

File: src/components/menus/common-contextual-menu.tsx (hunk 8 above)
Comment (Tristan-WorkGH) [reply src/components/menus/common-contextual-menu.tsx]: I had error of not using index in the key attribute.

File: src/components/custom-tree-item.tsx (hunk 9 above)
Comment (Tristan-WorkGH) [reply src/components/custom-tree-item.tsx:41]: Yes but not sure, there is a confusion with the classes system of the Tree components and the styles of Mui.

File: src/components/dialogs/export-case-dialog.tsx (hunk 10 above)
Comment (Tristan-WorkGH) [reply src/components/dialogs/export-case-dialog.tsx]: Types error: `fetchedFormats` is of `ExportFormats` type which is define as `type ExportFormats = { [p: string]: {/*...*/} } | []` so I need to known if it's a object before exploring it.

File: src/components/search/search-item.tsx (hunk 15 above)
Comment (Tristan-WorkGH) [reply src/components/search/search-item.tsx]: https://github.com/gridsuite/gridexplore-app/pull/546#discussion_r1816175769

**Review by dbraquart — COMMENTED** [review]
code ok
tests ok

just waiting for conflicts resolution

**Review by Meklo — APPROVED** [review]

**Review by dbraquart — APPROVED** [review]

**Pull request merged**